components/SerialServer/src/serial.c

```diff
@@ -21,90 +21,59 @@
 #include <platsupport/irq.h>
 #include "serial.h"
 #include "plat.h"
-#include "server_virtqueue.h"
 
 #define ESCAPE_CHAR '@'
 #define MAX_CLIENTS 12
-#define CLIENT_OUTPUT_BUFFER_SIZE 4096
-
-/* TODO: have the MultiSharedData template generate a header with these */
-void getchar_emit(unsigned int id) WEAK;
-seL4_Word getchar_enumerate_badge(unsigned int id) WEAK;
-unsigned int getchar_num_badges() WEAK;
-void *getchar_buf(unsigned int id) WEAK;
-void *processed_batch_buf(unsigned int id) WEAK;
-void *raw_batch_buf(unsigned int id) WEAK;
-int getchar_largest_badge(void) WEAK;
-
-typedef struct getchar_buffer {
+
+/* 4088 because the serial_shmem_t struct has to be 0x1000 bytes big */
+#define BUFSIZE 4088
+
+/* This struct occupies exactly 1 page and represents the data in the shmem region */
+typedef struct serial_shmem {
     uint32_t head;
     uint32_t tail;
-    char buf[4096 - 8];
-} client_buffer_t;
+    char buf[BUFSIZE];
+} serial_shmem_t;
+compile_time_assert(serial_shmem_1k_size, sizeof(serial_shmem_t) == 0x1000);
 
-compile_time_assert(getchar_buf_sized, sizeof(client_buffer_t) == sizeof(Buf));
+typedef struct client {
+    void (*notify)(void);
+    volatile serial_shmem_t *recv_buf;
+    volatile serial_shmem_t *send_buf;
```

Comment:
You could define the `output_buffer` here and `output_buffers_used` here, so it's all parts of the client context? Then any handler would resolve the client fist and everything else happen on the context.